### PR TITLE
Add java to the environment to bypass powerplantmatching hidden dependency

### LINKED ISSUE #367: FileNotFoundError when running build_powerplants
## Checklist

- [X] I am using the current [`main`](https://github.com/pypsa-meets-africa/pypsa-africa/tree/main) branch or the latest [release](https://github.com/pypsa-meets-africa/pypsa-africa/releases). Please indicate.
- [X] I am running on an up-to-date [`pypsa-africa` environment](https://github.com/pypsa-meets-africa/pypsa-africa/blob/main/envs/environment.yaml). Update via `conda env update -f envs/environment.yaml`.

## Describe the Bug

An attempt to run `build_powerplants` on a new machine has resulted in a `FileNotFoundError` with an error messages referencing a `"linkfile.txt"` in a temporary folder. It looked very much similar to the [problem](https://github.com/pypsa-meets-africa/pypsa-earth-sec/issues/103#issuecomment-1138367988) reported in the Workflow-check of the PyPSA-Earth-Sec model (@energyLS, many thanks for reporting it!) and was fixed using @Hazem-IEG suggestion by installing java:

`conda install -c bioconda java-jdk`

or

`conda install openjdk`

As a nice side effect, installing java could have resolved also some troubles with glpk which I have initially experienced in a fresh workspace. (But not sure if this is a reproducible)

## Error Message

The whole error message looked like:

```bash
INFO:snakemake.logging:
[Tue May 31 22:48:06 2022]
INFO:snakemake.logging:[Tue May 31 22:48:06 2022]
rule build_powerplants:
    input: networks/base.nc, configs/powerplantmatching_config.yaml, data/custom_powerplants.csv, data/clean/africa_all_generators.csv
    output: resources/powerplants.csv, resources/powerplants_osm2pm.csv
    log: logs/build_powerplants.log
    jobid: 0
    reason: Missing output files: resources/powerplants.csv, resources/powerplants_osm2pm.csv
    resources: tmpdir=/var/folders/qn/vpndfm21795ckkq89np1ckp40000gn/T, mem=500
INFO:snakemake.logging:rule build_powerplants:
    input: networks/base.nc, configs/powerplantmatching_config.yaml, data/custom_powerplants.csv, data/clean/africa_all_generators.csv
    output: resources/powerplants.csv, resources/powerplants_osm2pm.csv
    log: logs/build_powerplants.log
    jobid: 0
    reason: Missing output files: resources/powerplants.csv, resources/powerplants_osm2pm.csv
    resources: tmpdir=/var/folders/qn/vpndfm21795ckkq89np1ckp40000gn/T, mem=500

INFO:snakemake.logging:
INFO:pypsa.io:Imported network base.nc has buses, lines, links
INFO:powerplantmatching.collection:Create combined dataset for GEO, GPD
INFO:powerplantmatching.cleaning:Aggregating blocks in data source 'GEO'.
Traceback (most recent call last):
  File "~/Documents/_github_/pypsa-africa/.snakemake/scripts/tmpikj0kau5.build_powerplants.py", line 260, in <module>
    pm.powerplants(from_url=False, update=True, config=config)
  File "~/opt/miniconda3/envs/pypsa-africa/lib/python3.10/site-packages/powerplantmatching/collection.py", line 223, in matched_data
    matched = collect(matching_sources, config=config, **collection_kwargs)
  File "~/opt/miniconda3/envs/pypsa-africa/lib/python3.10/site-packages/powerplantmatching/collection.py", line 96, in collect
    dfs = parmap(df_by_name, datasets)
  File "~/opt/miniconda3/envs/pypsa-africa/lib/python3.10/site-packages/powerplantmatching/utils.py", line 378, in parmap
    return list(map(f, arg_list))
  File "~/opt/miniconda3/envs/pypsa-africa/lib/python3.10/site-packages/powerplantmatching/collection.py", line 74, in df_by_name
    return aggregate_units(df, dataset_name=name, config=config)
  File "~/opt/miniconda3/envs/pypsa-africa/lib/python3.10/site-packages/powerplantmatching/cleaning.py", line 447, in aggregate_units
    duplicates = pd.concat([duke(df.query("Country == @c")) for c in countries])
  File "~/opt/miniconda3/envs/pypsa-africa/lib/python3.10/site-packages/powerplantmatching/cleaning.py", line 447, in <listcomp>
    duplicates = pd.concat([duke(df.query("Country == @c")) for c in countries])
  File "~/opt/miniconda3/envs/pypsa-africa/lib/python3.10/site-packages/powerplantmatching/duke.py", line 152, in duke
    return pd.read_csv(
  File "~/opt/miniconda3/envs/pypsa-africa/lib/python3.10/site-packages/pandas/util/_decorators.py", line 311, in wrapper
    return func(*args, **kwargs)
  File "~/opt/miniconda3/envs/pypsa-africa/lib/python3.10/site-packages/pandas/io/parsers/readers.py", line 680, in read_csv
    return _read(filepath_or_buffer, kwds)
  File "~/opt/miniconda3/envs/pypsa-africa/lib/python3.10/site-packages/pandas/io/parsers/readers.py", line 575, in _read
    parser = TextFileReader(filepath_or_buffer, **kwds)
  File "~/opt/miniconda3/envs/pypsa-africa/lib/python3.10/site-packages/pandas/io/parsers/readers.py", line 933, in __init__
    self._engine = self._make_engine(f, self.engine)
  File "~/opt/miniconda3/envs/pypsa-africa/lib/python3.10/site-packages/pandas/io/parsers/readers.py", line 1217, in _make_engine
    self.handles = get_handle(  # type: ignore[call-overload]
  File "~/opt/miniconda3/envs/pypsa-africa/lib/python3.10/site-packages/pandas/io/common.py", line 789, in get_handle
    handle = open(
FileNotFoundError: [Errno 2] No such file or directory: '/var/folders/qn/vpndfm21795ckkq89np1ckp40000gn/T/tmpi9oa1sgi/linkfile.txt'
[Tue May 31 22:48:13 2022]
INFO:snakemake.logging:[Tue May 31 22:48:13 2022]
Error in rule build_powerplants:
    jobid: 0
    output: resources/powerplants.csv, resources/powerplants_osm2pm.csv
    log: logs/build_powerplants.log (check log file(s) for error message)

ERROR:snakemake.logging:Error in rule build_powerplants:
    jobid: 0
    output: resources/powerplants.csv, resources/powerplants_osm2pm.csv
    log: logs/build_powerplants.log (check log file(s) for error message)

RuleException:
CalledProcessError in line 333 of ~/Documents/_github_/pypsa-africa/Snakefile:
Command 'set -euo pipefail;  ~/opt/miniconda3/envs/pypsa-africa/bin/python3.10 ~/Documents/_github_/pypsa-africa/.snakemake/scripts/tmpikj0kau5.build_powerplants.py' returned non-zero exit status 1.
  File "~/Documents/_github_/pypsa-africa/Snakefile", line 333, in __rule_build_powerplants
  File "~/opt/miniconda3/envs/pypsa-africa/lib/python3.10/concurrent/futures/thread.py", line 58, in run
ERROR:snakemake.logging:RuleException:
CalledProcessError in line 333 of ~/Documents/_github_/pypsa-africa/Snakefile:
Command 'set -euo pipefail;  ~/opt/miniconda3/envs/pypsa-africa/bin/python3.10 ~/Documents/_github_/pypsa-africa/.snakemake/scripts/tmpikj0kau5.build_powerplants.py' returned non-zero exit status 1.
  File "~/Documents/_github_/pypsa-africa/Snakefile", line 333, in __rule_build_powerplants
  File "~/opt/miniconda3/envs/pypsa-africa/lib/python3.10/concurrent/futures/thread.py", line 58, in run
Removing output files of failed job build_powerplants since they might be corrupted:
resources/powerplants_osm2pm.csv
WARNING:snakemake.logging:Removing output files of failed job build_powerplants since they might be corrupted:
resources/powerplants_osm2pm.csv
Shutting down, this might take some time.
```

### PR DESCRIPTION
**Pull request by davide-f**
# Closes #367 

## Changes proposed in this Pull Request
Add java dependency to pypsa-africa environment to solve the issue with powerplantmatching

## Checklist

- [x] I tested my contribution locally and it seems to work fine.
- [x] Code and workflow changes are sufficiently documented.
- [x] Newly introduced dependencies are added to `envs/environment.yaml` and `envs/environment.docs.yaml`.
- [x] Changes in configuration options are added in all of `config.default.yaml`, `config.tutorial.yaml`, and `test/config.test1.yaml`.
- [x] Changes in configuration options are also documented in `doc/configtables/*.csv` and line references are adjusted in `doc/configuration.rst` and `doc/tutorial.rst`.
- [x] A note for the release notes `doc/release_notes.rst` is amended in the format of previous release notes, including reference to the requested PR.

### CODE REVIEW

**Comment by davide-f** [comment]
@pz-max install-jdk may not install java but an utility that helps installing java ( https://pypi.org/project/install-jdk/ ).
I've searched if it is possible to install it with pip, but it seems not so trivial https://stackoverflow.com/questions/71181609/how-to-install-java-jdk-jre-with-pip

I think we can install openjdk with our environment and we bypass this problem; that hidden dependency is also known in powerplantmatching as the final solution will be available we will go for it

@ekatef this may be of interest

**Comment by pz-max** [comment]
Experiencing at the moment the CI error @davide-f 
Over the conda-forge it will load `openjdk=17.0.3`
Trying to downgrade to see if something works

**Comment by davide-f** [comment]
> Experiencing at the moment the CI error @davide-f Over the conda-forge it will load `openjdk=17.0.3` Trying to downgrade to see if something works

The issue is a bit tricky: in mac and windows there are issues of versioning: two java versions are installed at the same time and there must be conflicts about "which one" is chosen.
Testing

**Comment by ekatef** [comment]
I have had an impression that `install-jdk` should run a Github action (which is something makefile-like, if I understand properly) when being included into `setup.py`... 

@davide-f, thank you so much for taking care of it! Actually, that's one of my nightmares when package installation fails due to some Java troubles :)

Is it not possible to check if JRK is still installed on the system before trying to install `openjdk`?

**Comment by davide-f** [comment]
> I have had an impression that `install-jdk` should run a Github action (which is something makefile-like, if I understand properly) when being included into `setup.py`...
> 
> @davide-f, thank you so much for taking care of it! Actually, that's one of my nightmares when package installation fails due to some Java troubles :)
> 
> Is it not possible to check if JRK is still installed on the system before trying to install `openjdk`?

The fact is not the CI but a user: I feel like currently trying to fix this issue with conda seems a nightmare and not really worth spending the time that is requiring.
In my opinion, the best approach would be to specify in the installation guide to make sure to install java in their OS.
If we put java in the environment, it must work for everyone, otherwise we better not include it.
I've found various tricks to make java accessible, like specifying a custom environment variable, but they seem tricks leading to low maintenance.
By adding the dependency that may be pypassed. Probably, in a long-term perspective, ppl may get rid of the java dependency, though it requires some changes

In short my proposal is:
- keep the environment as is in main
- change the documentation and installation guide to make sure to install java on their OS

@pz-max if you also agree, this PR can be closed or modified so as to change the installation guide

**Comment by ekatef** [comment]
> In short my proposal is:
> 
> keep the environment as is in main
> change the documentation and installation guide to make sure to install java on their OS

As a user I would be quite happy with this approach :)

**Comment by pz-max** [comment]
Something that solve the issue for the meantime:
- Download and install Java from here https://www.oracle.com/java/technologies/downloads/
- Close and open VScode
- Test if Java works in the cmd terminal `java --version`

**Comment by davide-f** [comment]
> Something that solve the issue for the meantime:
> 
> * Download and install Java from here https://www.oracle.com/java/technologies/downloads/
> * Close and open VScode
> * Test if Java works in the cmd terminal `java --version`

I'd suggest to add into the installation guide to make sure to have it installed; I can revise this pr accordingly
